beacon-chain/sync/initial-sync/round_robin.go

```diff
@@ -0,0 +1,258 @@
+				fmt.Sprintf("%d/%d", len(peers), len(peerstatus.Keys())),
+			).WithField(
+				"blocks per second",
+				fmt.Sprintf("%.1f", rate),
+			).Infof("Processing block %d/%d. Estimated %.0f seconds remaining.", blk.Slot, slotsSinceGenesis(genesis), timeRemaining)
+			if err := s.chain.ReceiveBlockNoPubsubForkchoice(ctx, blk); err != nil {
+				return err
+			}
+		}
+	}
+
+	log.Debug("Synced to finalized epoch. Syncing blocks to head slot now.")
+
+	if s.chain.HeadSlot() == slotsSinceGenesis(genesis) {
+		return nil
+	}
+
+	// Step 2 - sync to head from any single peer.
+	// This step might need to be improved for cases where there has been a long period since
+	// finality. This step is less important than syncing to finality in terms of threat
+	// mitigation. We are already convinced that we are on the correct finalized chain. Any blocks
+	// we receive there after must build on the finalized chain or be considered invalid during
+	// fork choice resolution / block processing.
+	best := bestPeer()
+	root, _, _ := highestFinalized()
+	req := &p2ppb.BeaconBlocksByRangeRequest{
+		HeadBlockRoot: root,
+		StartSlot:     s.chain.HeadSlot() + 1,
+		Count:         slotsSinceGenesis(genesis) - s.chain.HeadSlot() + 1,
+		Step:          1,
+	}
+
+	log.WithField("req", req).WithField("peer", best.Pretty()).Debug(
+		"Sending batch block request",
+	)
+
+	resp, err := s.requestBlocks(ctx, req, best)
+	if err != nil {
+		return err
+	}
+
+	for _, blk := range resp {
+		if err := s.chain.ReceiveBlockNoPubsubForkchoice(ctx, blk); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *InitialSync) requestBlocks(ctx context.Context, req *p2ppb.BeaconBlocksByRangeRequest, pid peer.ID) ([]*eth.BeaconBlock, error) {
+	log.WithField("peer", pid.Pretty()).WithField("req", req).Debug("requesting blocks")
+	stream, err := s.p2p.Send(ctx, req, pid)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to send request to peer")
+	}
+	defer stream.Close()
+
+	resp := make([]*eth.BeaconBlock, 0, req.Count)
+	for {
+		blk, err := prysmsync.ReadChunkedBlock(stream, s.p2p)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to read chunked block")
+		}
+		resp = append(resp, blk)
+	}
+
+	return resp, nil
+}
+
+// highestFinalizedEpoch as reported by peers. This is the absolute highest finalized epoch as
+// reported by peers.
+func highestFinalizedEpoch() uint64 {
+	var finalizedEpoch uint64
+	for _, k := range peerstatus.Keys() {
+		s := peerstatus.Get(k)
+		finalizedEpoch = mathutil.Max(s.FinalizedEpoch, finalizedEpoch)
+	}
+
+	return finalizedEpoch
+}
+
+// highestFinalized returns the highest finalized epoch that is agreed upon by the majority of
+// peers. This method may not return the absolute highest finalized, but the finalized epoch in
+// which most peers can serve blocks. Ideally, all peers would be reporting the same finalized
+// epoch.
+func highestFinalized() ([]byte, uint64, []peer.ID) {
```

Comment:
What's the difference between this func and the one above it? The naming/comments are very confusing here

Reply:
Tried to add some clarification, let me know what you think